In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

In [ ]:
data_csvfile = pd.read_csv('lab6_ex1.csv')

In [ ]:
data_csvfile.index

RangeIndex(start=0, stop=9, step=1)

In [ ]:
print("Number of rows: ", len(data_csvfile.index))

Number of rows:  9


In [ ]:
#to get the row contents corresponding to Ahmedabad
print(data_csvfile.iloc[0, :])

Warehouses    Ahmedabad
Hassan              827
Jodhpur             817
Trichy             1470
Simla              2272
Vijayawada         1615
Dibru              3143
Raipur             1264
Agartala           4465
Stock               350
Name: 0, dtype: object


In [ ]:
#to get the column contents corresponding to Hassan
print(data_csvfile.iloc[:, 1])

0     827
1     559
2    1109
3    2943
4    3742
5    2353
6    1356
7     195
8     205
Name: Hassan, dtype: int64


In [ ]:
#to get the row contents corresponding to Mumbai
print(data_csvfile.iloc[6, :])

Warehouses    Mumbai
Hassan          1356
Jodhpur         1617
Trichy          2045
Simla           2941
Vijayawada      2334
Dibru           4151
Raipur          1937
Agartala        4287
Stock            325
Name: 6, dtype: object


In [ ]:
data_csvfile.columns

Index(['Warehouses', 'Hassan', 'Jodhpur', 'Trichy', 'Simla', 'Vijayawada',
       'Dibru', 'Raipur', 'Agartala', 'Stock'],
      dtype='object')

In [ ]:
#create a model
model_lab6_ex1 = ConcreteModel()

In [ ]:
#setting the values using python variables
M = len(data_csvfile.index) - 1
print("Number of Warehouses: ", M)

Number of Warehouses:  8


In [ ]:
#print the number of cities
N = len(data_csvfile.columns) - 2
print("Number of cities: ", N)

Number of cities:  8


In [ ]:
row_indices = range(M)
col_indices =  range(N)
print(row_indices)
print(col_indices)

range(0, 8)
range(0, 8)


In [ ]:
#get contents of cost
for row_idx in row_indices:
  for col_idx in col_indices:
    print('Cost[', row_idx, '][', col_idx + 1, ']: ', data_csvfile.iloc[row_idx, col_idx + 1])

Cost[ 0 ][ 1 ]:  827
Cost[ 0 ][ 2 ]:  817
Cost[ 0 ][ 3 ]:  1470
Cost[ 0 ][ 4 ]:  2272
Cost[ 0 ][ 5 ]:  1615
Cost[ 0 ][ 6 ]:  3143
Cost[ 0 ][ 7 ]:  1264
Cost[ 0 ][ 8 ]:  4465
Cost[ 1 ][ 1 ]:  559
Cost[ 1 ][ 2 ]:  1832
Cost[ 1 ][ 3 ]:  751
Cost[ 1 ][ 4 ]:  3772
Cost[ 1 ][ 5 ]:  637
Cost[ 1 ][ 6 ]:  3773
Cost[ 1 ][ 7 ]:  1856
Cost[ 1 ][ 8 ]:  5459
Cost[ 2 ][ 1 ]:  1109
Cost[ 2 ][ 2 ]:  2443
Cost[ 2 ][ 3 ]:  577
Cost[ 2 ][ 4 ]:  3865
Cost[ 2 ][ 5 ]:  798
Cost[ 2 ][ 6 ]:  3521
Cost[ 2 ][ 7 ]:  1984
Cost[ 2 ][ 8 ]:  5303
Cost[ 3 ][ 1 ]:  2943
Cost[ 3 ][ 2 ]:  1471
Cost[ 3 ][ 3 ]:  4676
Cost[ 3 ][ 4 ]:  1337
Cost[ 3 ][ 5 ]:  2702
Cost[ 3 ][ 6 ]:  2741
Cost[ 3 ][ 7 ]:  2482
Cost[ 3 ][ 8 ]:  4148
Cost[ 4 ][ 1 ]:  3742
Cost[ 4 ][ 2 ]:  3089
Cost[ 4 ][ 3 ]:  2931
Cost[ 4 ][ 4 ]:  3199
Cost[ 4 ][ 5 ]:  2226
Cost[ 4 ][ 6 ]:  1739
Cost[ 4 ][ 7 ]:  1691
Cost[ 4 ][ 8 ]:  2128
Cost[ 5 ][ 1 ]:  2353
Cost[ 5 ][ 2 ]:  1160
Cost[ 5 ][ 3 ]:  3295
Cost[ 5 ][ 4 ]:  368
Cost[ 5 ][ 5 ]:  2369
Cost[ 5 ][ 6 ]:  2

In [ ]:
#declare the decision variables in the model
model_lab6_ex1.x = Var(row_indices, col_indices, domain = NonNegativeIntegers)

In [ ]:
model_lab6_ex1.objective = Objective(expr = sum(data_csvfile.iloc[i, j+1]*model_lab6_ex1.x[i, j] for i in row_indices for j in col_indices), sense = minimize)

In [ ]:
model_lab6_ex1.constraints = ConstraintList()

In [ ]:
#to get the supply limit at warehouses
print(data_csvfile.iloc[:8, N+1])

0    350
1    185
2    305
3    180
4    205
5    175
6    325
7    370
Name: Stock, dtype: int64


In [ ]:
#to get the demands at cities
print(data_csvfile.iloc[M, 1:N+1])

Hassan        205
Jodhpur       300
Trichy        425
Simla         270
Vijayawada    375
Dibru         225
Raipur        185
Agartala      110
Name: 8, dtype: object


In [ ]:
#first we will add the supply limit constraints
#supply constraint for Ahmedabad
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[0, j] for j in col_indices) <= data_csvfile.iloc[0, N+1])

#supply constraint for Bengaluru
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[1, j] for j in col_indices) <= data_csvfile.iloc[1, N+1])

#supply constraint for Chennai
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[2, j] for j in col_indices) <= data_csvfile.iloc[2, N+1])

#supply constraint for Delhi
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[3, j] for j in col_indices) <= data_csvfile.iloc[3, N+1])

#supply constraint for Kolkata
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[4, j] for j in col_indices) <= data_csvfile.iloc[4, N+1])

#supply constraint for Lucknow
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[5, j] for j in col_indices) <= data_csvfile.iloc[5, N+1])

#supply constraint for Mumbai
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[6, j] for j in col_indices) <= data_csvfile.iloc[6, N+1])

#supply constraint for Hyderabad
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[7, j] for j in col_indices) <= data_csvfile.iloc[7, N+1])



In [ ]:
#now we will add the demand constraints
#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 0] for i in row_indices) == data_csvfile.iloc[M, 1]) 

#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 1] for i in row_indices) == data_csvfile.iloc[M, 2]) 

#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 2] for i in row_indices) == data_csvfile.iloc[M, 3])

#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 3] for i in row_indices) == data_csvfile.iloc[M, 4]) 

#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 4] for i in row_indices) == data_csvfile.iloc[M, 5]) 

#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 5] for i in row_indices) == data_csvfile.iloc[M, 6]) 

#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 6] for i in row_indices) == data_csvfile.iloc[M, 7]) 

#demand constraint for Hassan
model_lab6_ex1.constraints.add(expr = sum(model_lab6_ex1.x[i, 7] for i in row_indices) == data_csvfile.iloc[M, 8]) 


In [ ]:
model_lab6_ex1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
    

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print("Solver Status: ", result.solver.status)
print("Solver Termination Condition", result.solver.termination_condition)

Solver Status:  ok
Solver Termination Condition optimal


In [ ]:
#display solution
print("\nObjective = ", model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print("Number of truck-loads of EVs from", data_csvfile['Warehouses'][i], "to", data_csvfile.columns[j + 1], ":", model_lab6_ex1.x[i, j].value)

print("n\Constraints")
model_lab6_ex1.constraints.display()


Objective =  2192870.0

Decision Variables
Number of truck-loads of EVs from Ahmedabad to Hassan : 0.0
Number of truck-loads of EVs from Ahmedabad to Jodhpur : 300.0
Number of truck-loads of EVs from Ahmedabad to Trichy : 0.0
Number of truck-loads of EVs from Ahmedabad to Simla : 0.0
Number of truck-loads of EVs from Ahmedabad to Vijayawada : 0.0
Number of truck-loads of EVs from Ahmedabad to Dibru : 45.0
Number of truck-loads of EVs from Ahmedabad to Raipur : 5.0
Number of truck-loads of EVs from Ahmedabad to Agartala : 0.0
Number of truck-loads of EVs from Bengal to Hassan : 0.0
Number of truck-loads of EVs from Bengal to Jodhpur : 0.0
Number of truck-loads of EVs from Bengal to Trichy : 120.0
Number of truck-loads of EVs from Bengal to Simla : 0.0
Number of truck-loads of EVs from Bengal to Vijayawada : 65.0
Number of truck-loads of EVs from Bengal to Dibru : 0.0
Number of truck-loads of EVs from Bengal to Raipur : 0.0
Number of truck-loads of EVs from Bengal to Agartala : 0.0
Numb

In [ ]:
data_csvfile.iloc[5, 4] = 10000000

In [ ]:
model_lab6_ex1.objective = Objective(expr = sum(data_csvfile.iloc[i, j+1]*model_lab6_ex1.x[i, j] for i in row_indices for j in col_indices), sense = minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print("Solver Status: ", result.solver.status)
print("Solver Termination Condition", result.solver.termination_condition)

Solver Status:  ok
Solver Termination Condition optimal


In [ ]:
#display solution
print("\nObjective = ", model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print("Number of truck-loads of EVs from", data_csvfile['Warehouses'][i], "to", data_csvfile.columns[j + 1], ":", model_lab6_ex1.x[i, j].value)

print("n\Constraints")
model_lab6_ex1.constraints.display()


Objective =  2380570.0

Decision Variables
Number of truck-loads of EVs from Ahmedabad to Hassan : 0.0
Number of truck-loads of EVs from Ahmedabad to Jodhpur : 300.0
Number of truck-loads of EVs from Ahmedabad to Trichy : 0.0
Number of truck-loads of EVs from Ahmedabad to Simla : 0.0
Number of truck-loads of EVs from Ahmedabad to Vijayawada : 0.0
Number of truck-loads of EVs from Ahmedabad to Dibru : 0.0
Number of truck-loads of EVs from Ahmedabad to Raipur : 50.0
Number of truck-loads of EVs from Ahmedabad to Agartala : 0.0
Number of truck-loads of EVs from Bengal to Hassan : 0.0
Number of truck-loads of EVs from Bengal to Jodhpur : 0.0
Number of truck-loads of EVs from Bengal to Trichy : 120.0
Number of truck-loads of EVs from Bengal to Simla : 0.0
Number of truck-loads of EVs from Bengal to Vijayawada : 65.0
Number of truck-loads of EVs from Bengal to Dibru : 0.0
Number of truck-loads of EVs from Bengal to Raipur : 0.0
Number of truck-loads of EVs from Bengal to Agartala : 0.0
Numb

In [ ]:
data_csvfile.iloc[4, 6] = 100000
data_csvfile.iloc[3, 7] = 200000

In [ ]:
model_lab6_ex1.objective = Objective(expr = sum(data_csvfile.iloc[i, j+1]*model_lab6_ex1.x[i, j] for i in row_indices for j in col_indices), sense = minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print("Solver Status: ", result.solver.status)
print("Solver Termination Condition", result.solver.termination_condition)

Solver Status:  ok
Solver Termination Condition optimal


In [ ]:
#display solution
print("\nOnbjective = ", model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print("Number of truck-loads of EVs from", data_csvfile['Warehouses'][i], "to", data_csvfile.columns[j + 1], ":", model_lab6_ex1.x[i, j].value)

print("n\Constraints")
model_lab6_ex1.constraints.display()


Onbjective =  2518200.0

Decision Variables
Number of truck-loads of EVs from Ahmedabad to Hassan : 0.0
Number of truck-loads of EVs from Ahmedabad to Jodhpur : 300.0
Number of truck-loads of EVs from Ahmedabad to Trichy : 0.0
Number of truck-loads of EVs from Ahmedabad to Simla : 0.0
Number of truck-loads of EVs from Ahmedabad to Vijayawada : 0.0
Number of truck-loads of EVs from Ahmedabad to Dibru : 50.0
Number of truck-loads of EVs from Ahmedabad to Raipur : 0.0
Number of truck-loads of EVs from Ahmedabad to Agartala : 0.0
Number of truck-loads of EVs from Bengal to Hassan : 0.0
Number of truck-loads of EVs from Bengal to Jodhpur : 0.0
Number of truck-loads of EVs from Bengal to Trichy : 120.0
Number of truck-loads of EVs from Bengal to Simla : 0.0
Number of truck-loads of EVs from Bengal to Vijayawada : 65.0
Number of truck-loads of EVs from Bengal to Dibru : 0.0
Number of truck-loads of EVs from Bengal to Raipur : 0.0
Number of truck-loads of EVs from Bengal to Agartala : 0.0
Num

**Report on exercise 1 part 1:**

The LP formulation is as follows:

$ min \sum_{w=1}^8\sum_{m=1}^8 C(w,m)*x_{wm} $

$s.t. \sum_{w=1}^8x_{wm}=d_m   ∀  mϵ${ 1,2,....8 }

$ \sum_{m=1}^8x_{wm} <= s_w ∀wϵ${1,2,....8}

$ x_{wm}>= 0  ∀w,m ϵ$ {1,2,....,8}

where C(w, m) is the cost of transporting one unit of truck-load of EVs from w to m and $x_{wm}$ is the number of units of truck-loads transferring from w to m. 


**Report on Exercise 1 part 7:**

**Total Cost:**
2192870.0

**Number of truck-loads of EVs from warehouses to markets are as follows:**

Number of truck-loads of EVs from Ahmedabad to Jodhpur : 300.0


Number of truck-loads of EVs from Ahmedabad to Dibru : 45.0

Number of truck-loads of EVs from Ahmedabad to Raipur : 5.0


Number of truck-loads of EVs from Bengal to Trichy : 120.0

Number of truck-loads of EVs from Bengal to Vijayawada : 65.0

Number of truck-loads of EVs from Chennai to Trichy : 305.0


Number of truck-loads of EVs from Delhi to Simla : 95.0

Number of truck-loads of EVs from Delhi to Dibru : 85.0

Number of truck-loads of EVs from Kolkata to Dibru : 95.0


Number of truck-loads of EVs from Kolkata to Agartala : 110.0


Number of truck-loads of EVs from Lucknow to Simla : 175.0

Number of truck-loads of EVs from Mumbai to Hassan : 205.0

Number of truck-loads of EVs from Mumbai to Raipur : 120.0

Number of truck-loads of EVs from Hyderabad to Vijayawada : 310.0

Number of truck-loads of EVs from Hyderabad to Raipur : 60.0



**Report on Exercise 1 part 8:**

Now, Lucknow-Simla link is disrupted that means Lucknow is now not supplying to Simla so we'll change the cost value in the data file. We'll increase the cost (because it's a minimization problem) for Lucknow-Simla link so that Lucknow-Simla supply become zero.

**Report on Exercise 1 part 9:**

Solution to the new problem is:

**Objective** =  2380570.0

**Decision Variables:**

Number of truck-loads of EVs from Ahmedabad to Jodhpur : 300.0

Number of truck-loads of EVs from Ahmedabad to Raipur : 50.0

Number of truck-loads of EVs from Bengal to Trichy : 120.0

Number of truck-loads of EVs from Bengal to Vijayawada : 65.0

Number of truck-loads of EVs from Chennai to Trichy : 305.0

Number of truck-loads of EVs from Delhi to Simla : 180.0

Number of truck-loads of EVs from Kolkata to Dibru : 95.0

Number of truck-loads of EVs from Kolkata to Agartala : 110.0

Number of truck-loads of EVs from Lucknow to Dibru : 130.0

Number of truck-loads of EVs from Lucknow to Raipur : 45.0

Number of truck-loads of EVs from Mumbai to Hassan : 205.0

Number of truck-loads of EVs from Mumbai to Simla : 90.0

Number of truck-loads of EVs from Mumbai to Raipur : 30.0

Number of truck-loads of EVs from Hyderabad to Vijayawada : 310.0

Number of truck-loads of EVs from Hyderabad to Raipur : 60.0

**Constraints:**

constraints : Size=16

    Key : Lower : Body  : Upper
      1 :  None : 350.0 : 350.0
      2 :  None : 185.0 : 185.0
      3 :  None : 305.0 : 305.0
      4 :  None : 180.0 : 180.0
      5 :  None : 205.0 : 205.0
      6 :  None : 175.0 : 175.0
      7 :  None : 325.0 : 325.0
      8 :  None : 370.0 : 370.0
      9 : 205.0 : 205.0 : 205.0
     10 : 300.0 : 300.0 : 300.0
     11 : 425.0 : 425.0 : 425.0
     12 : 270.0 : 270.0 : 270.0
     13 : 375.0 : 375.0 : 375.0
     14 : 225.0 : 225.0 : 225.0
     15 : 185.0 : 185.0 : 185.0
     16 : 110.0 : 110.0 : 110.0


**Observations on result:**

1) Lucknow-Simla supply is zero after disruption.

2) Initially, Lucknow was supplying 175 electric vehicles to Simla but after disruption of Lucknow-Simla link, Lucknow is supplying 130 vehicles to **Dibrugarh** and 45 vehicles to **Raipur** because it has 175 available stocks. 

3) The demand of simla is 270, so it will now be fulfilled by other cities(Mumbai and Delhi).

4) Also, the supply from other cities changed. For example, initially supply from Ahmedabad to Dibru was 45 and to Raipur it was 5 but after disruption Ahmedabad stopped supplying to Dibru instead it started supplying more to Raipur.

5) The total cost increased.

**Report on Exercise 1 part 10:**

The new solution after disruption of links between Kolkata-Dibrugarh and Delhi-Raipur in addition to Lucknow-Simla is as follows:

**Objective** =  2518200.0

**Decision Variables:**

Number of truck-loads of EVs from Ahmedabad to Jodhpur : 300.0

Number of truck-loads of EVs from Ahmedabad to Dibru : 50.0

Number of truck-loads of EVs from Bengal to Trichy : 120.0

Number of truck-loads of EVs from Bengal to Vijayawada : 65.0

Number of truck-loads of EVs from Chennai to Trichy : 305.0

Number of truck-loads of EVs from Delhi to Simla : 180.0

Number of truck-loads of EVs from Kolkata to Raipur : 95.0

Number of truck-loads of EVs from Kolkata to Agartala : 110.0

Number of truck-loads of EVs from Lucknow to Dibru : 175.0

Number of truck-loads of EVs from Mumbai to Hassan : 205.0

Number of truck-loads of EVs from Mumbai to Simla : 90.0

Number of truck-loads of EVs from Mumbai to Raipur : 30.0

Number of truck-loads of EVs from Hyderabad to Vijayawada : 310.0

Number of truck-loads of EVs from Hyderabad to Raipur : 60.0

**Constraints**

constraints : Size=16

    Key : Lower : Body  : Upper
      1 :  None : 350.0 : 350.0
      2 :  None : 185.0 : 185.0
      3 :  None : 305.0 : 305.0
      4 :  None : 180.0 : 180.0
      5 :  None : 205.0 : 205.0
      6 :  None : 175.0 : 175.0
      7 :  None : 325.0 : 325.0
      8 :  None : 370.0 : 370.0
      9 : 205.0 : 205.0 : 205.0
     10 : 300.0 : 300.0 : 300.0
     11 : 425.0 : 425.0 : 425.0
     12 : 270.0 : 270.0 : 270.0
     13 : 375.0 : 375.0 : 375.0
     14 : 225.0 : 225.0 : 225.0
     15 : 185.0 : 185.0 : 185.0
     16 : 110.0 : 110.0 : 110.0

**Observations from the result:**

1) Total cost increased (even more than the last solution when only Lucknow-Simla link was disrupted).

2) Since, Kolkata-Dibrugarh link is disrupted, so kolkata started supplying to city Raipur, Demand of Dibrugarh is now fulfilled by city Lucknow and Ahmedabad.

3) Also, Delhi-Raipur link disrupted so, Delhi started supplying to Simla and cities Kolkata and Mumbai now fulfilling the remaining demand of Raipur.